Fix lesson preview indicator

alexsanford opened the pull request:
Closes #2137 

This change makes the Preview indicator for lessons consistent across Sensei.

Since there several different cases to address for this, I have provided a bit of an explanation below of each of the cases (after the testing instructions). Hopefully that will help with code review.

## Testing Instructions

- Create a course with a few lessons. Ensure that some lessons are in a module, and some are not. Check the "Allow this lesson to be viewed without purchase/login" checkbox in the "Preview" metabox for a few lessons: at least one in a module, and at least one that is not in a module.

- There are many cases to test. Please consider the following:

### Pages

The preview indicator should show up on the following pages:

- The Course page (`/course/<course-slug>`). The indicator should show up beside each lesson title. Please ensure that it is there for lessons with and without modules (the code is different for the two).

- The Lesson page (`/lesson/<lesson-slug>`). The indicator should show up beside the title at the top of the page. Please test with a preview lesson and a non-preview lesson. For the preview lesson, the pink notice should not appear (as it was replaced by the new preview indicator) but for the non-preview lesson the pink notice should still be there.

- The Lesson Tag page (`/lesson-tag/<tag-slug>`). The indicator should show up beside the title of the lesson. Ensure that the tag you are testing is set on both preview and non-preview lessons.

### When taking the course

Regardless of whether or not the course is linked to a WooCommerce product, when a user is taking a course, they should not see any preview indicators.

### Logged-out access

Regardless of the state of the Sensei setting "Users must be logged in to view Course and Lesson content", users should always see the preview indicators for courses if they are logged out, or logged in but not taking the course.

### WooCommerce

If the course is linked to a WooCommerce product, the preview indicators should all say "Free Preview" instead of "Preview". However, they should still appear in the same cases as before.

## Code Explanation

The code for the preview indicator was a bit scattered. I've created a helper method to create a consistent HTML tag, which all of the cases should be using now. See the method `sensei_lesson_preview_title_tag` in `includes/class-sensei-frontend.php`.

I've tweaked some CSS to make this indicator look good in all of the various cases. In short, I took the original CSS (which was in `modules-frontend.scss`) and moved it into `frontend/sensei.scss`. I then added a few tweaks for specific cases, building off of that core. See the two SCSS files for that.

The Lesson Page uses a hook on `the_title` to display the indicator using `sensei_lesson_preview_title` in `class-sensei-frontend.php`. That function now uses the helper described above.

On the Course Page, in the section that lists the lessons, there are two cases to consider: lessons with a module, and lessons without a module. For lessons with a module, these already had the appropriate indicator, although it always said "Free Preview" even when WooCommerce was not being used. I fixed this (it uses the new helper now) in `templates/single-course/modules.php`.

For lessons without a module, it uses a filter which called `the_lesson_meta` in `class-sensei-lesson.php`. This was updated to use the new helper. This function is also used on the Lesson Tag page.

The larger change in `class-sensei-lesson.php` is to remove the pink notice for preview lessons. These notices are handled by the `course_signup_link` method. I changed the method to return early for preview lessons, and removed the handling of preview lessons in the three specific cases.

It is worth noting that there is one more piece of code that appears to handle some preview indicator logic: the method [`sensei_course_preview_titles`](https://github.com/Automattic/sensei/blob/e38e907cbcb748352b497867e8e55081435fa7d5/includes/class-sensei-modules.php#L481) within `class-sensei-modules.php`. In my searches, it appeared that this function was only used in a single filter (`sensei_module_lesson_list_title`, added in [`includes/class-sensei-modules.php` line 101](https://github.com/Automattic/sensei/blob/e38e907cbcb748352b497867e8e55081435fa7d5/includes/class-sensei-modules.php#L101)), but this filter is never actually called within Sensei or its extensions. From a Github search, I see that it may be used by some 3rd-party extensions. But its usage is not clear to me, so I left it alone.

donnapep commented:
The _Free Preview_ label is a different size depending on which page you're on (course, lesson, tag). Could they all be the same size as on the single course page?

donnapep commented on templates/single-course/modules.php:
```diff
@@ -106,7 +106,7 @@
 	                                    if ( Sensei_Utils::is_preview_lesson( get_the_ID() ) && ! Sensei_Utils::user_started_course( $course_id, get_current_user_id() )  ) { ?>
 
-	                                        <span class="preview-label"><?php _e( 'Free Preview', 'woothemes-sensei' ); ?></span>
+	                                        <?php echo Sensei()->frontend->sensei_lesson_preview_title_tag( $course_id ); ?>
```
Might as well put this inside of the previous `<?php ?>` instead of creating another one.

alexsanford replied:
Done!

donnapep commented on includes/class-sensei-frontend.php:
```diff
@@ -1140,18 +1140,24 @@ public function sensei_lesson_meta( $post_id = 0 ) {
+    	    	$preview_text = __( 'Free Preview', 'woothemes-sensei' );
     	    } // End If Statement
     	}
     	return $preview_text;
```
Should we make this a filter?

donnapep commented on includes/class-sensei-frontend.php:
```diff
@@ -1140,18 +1140,24 @@ public function sensei_lesson_meta( $post_id = 0 ) {
+    	    	$preview_text = __( 'Free Preview', 'woothemes-sensei' );
     	    } // End If Statement
     	}
     	return $preview_text;
```
Should we add a filter here?

alexsanford replied:
Good idea. Done!

alexsanford commented:
> The Free Preview label is a different size depending on which page you're on (course, lesson, tag). Could they all be the same size as on the single course page?

This is a bit tricky, currently. The problem is that on some pages (e.g. Lesson, Tag) the label is inside a heading tag (`h1`, `h2`). Depending on the theme, that's going to make these ones different sizes from the rest.

What I did in the CSS was change the `h1` and `h2` versions of the label to be 75% the default font size, because I felt that it looked better that way when compared to the text in the actual header. On the other pages (e.g. Single Course) I left it at its default font size, because that seemed to look ok.

I'm thinking it's going to be hard to make them all a consistent size while also having them look good on many themes. For example, in some themes the headers are larger than regular paragraph text, in others they're smaller. Which size should we choose? And how do we set it up in CSS such that we're not hardcoding the size (which would be bad for theme compatibility)?

I think that the current solution is a decent tradeoff, but I'm open to suggestions! 😄

donnapep commented:
> This is a bit tricky, currently. The problem is that on some pages (e.g. Lesson, Tag) the label is inside a heading tag (h1, h2). Depending on the theme, that's going to make these ones different sizes from the rest.

What if we move it out of the heading tag and style it separately? That would eliminate the problem of inheriting styles from the parent `h1` or `h2`. I don't think there's a compelling reason to keep it in there.

donnapep commented:
> I've removed the method that filter used to use because it isn't used anywhere else. I can't think of any cases where that would be a problem, but it is technically not backwards compatible.

I've done this before and it ended up being my bad. Third party plugins have access to `public` functions, so we risk breaking those by removing any of them. See here for [reference](https://a8c.slack.com/archives/C6GGX896G/p1526927480000026). Not sure what it looks like to "stub" a method.

Looks like the lesson tag page needs fixing as the lesson title text exists outside of the `h2`.

Also, I hope you don't mind, but I played with some CSS in order to vertically align the preview label with the lesson title and to make its dimensions the same as those on the single course page (114.47 x 35.03 in Storefront). Here's the Sass I used (this may require some tweaking to make it also work on the lesson tag page):
```
header {
		display: flex;
		align-items: center;
		margin-bottom: 0.5em;

		.preview-label {
			margin-left: 10px;
			font-size: 1.1em;
		}

		&.lesson-title h1 {
			display: inline-block;
			margin: 0;
		}
}
```

alexsanford commented:
> Third party plugins have access to public functions, so we risk breaking those by removing any of them.

Good point. I added the method back in.

> Looks like the lesson tag page needs fixing as the lesson title text exists outside of the h2.

Yep, my mistake. Fixed now.

> I played with some CSS in order to vertically align the preview label with the lesson title and to make its dimensions the same as those on the single course page (114.47 x 35.03 in Storefront).

Looks great! I updated it, and just made one tweak for the lesson-tag page for flexbox to work properly there.
